fillers/eips/eip4844/datahash_opcode.py

```diff
@@ -538,118 +545,238 @@ def test_datahash_gas(_: str):
+    """
+    Tests that the `DATAHASH` opcode returns the correct versioned hash for
+    various valid indexes
+    """
+    NUM_BLOCKS = 10
+
+    pre = {
+        TestAddress: Account(balance=10000000000000000000000),
+    }
+    blocks = []
+    post = {}
+
+    datahash_verbatim = "verbatim_{}i_{}o".format(
+        Op.DATAHASH.popped_stack_items, Op.DATAHASH.pushed_stack_items
     )
 
-    post = {
-        datahash_zero_gas_measure_code_address: Account(
-            storage={
-                0: DATAHASH_GAS_COST,
-            }
-        ),
-        datahash_max_gas_measure_code_address: Account(
+    datahash_code = Yul(
+        f"""
+        {{
+            let pos := 0
+            let end := {MAX_BLOB_PER_BLOCK}
+            for {{}} lt(pos, end) {{ pos := add(pos, 1) }}
+            {{
+                let datahash := {datahash_verbatim}(
+                    hex"{Op.DATAHASH.hex()}",pos)
+                sstore(pos, datahash)
+            }}
+        }}
+        """
+    )
+
+    # Create a list of random blob hashes
+    blob_ver_hashes = [
+        to_hash_bytes(random.randint(MAX_BLOB_PER_BLOCK, 2**256 - 1))
+        for _ in range(MAX_BLOB_PER_BLOCK * NUM_BLOCKS)
+    ]
+
+    for i in range(NUM_BLOCKS):
+        address = to_address(0x100 + i * 0x100)
+        pre[address] = Account(code=datahash_code)
+        blocks.append(
+            Block(
+                txs=[  # Create tx with max blobs per block
+                    Transaction(
+                        ty=5,
+                        nonce=i,
+                        data=to_hash_bytes(0),
+                        to=address,
+                        gas_limit=3000000,
+                        max_fee_per_gas=10,
+                        max_priority_fee_per_gas=10,
+                        max_fee_per_data_gas=10,
+                        access_list=[],
+                        blob_versioned_hashes=blob_ver_hashes[
+                            (i * MAX_BLOB_PER_BLOCK) : (i + 1)
+                            * MAX_BLOB_PER_BLOCK
+                        ],
+                    )
+                ]
+            )
+        )
+        post[address] = Account(
             storage={
-                0: DATAHASH_GAS_COST,
+                index: blob_ver_hashes[i * MAX_BLOB_PER_BLOCK + index]
+                for index in range(MAX_BLOB_PER_BLOCK)
             }
-        ),
-    }
+        )
 
-    yield StateTest(
-        env=env,
+    yield BlockchainTest(
         pre=pre,
         post=post,
-        txs=[tx1, tx2],
-        tag="cost_tx_type_5",
+        blocks=blocks,
     )
 
-    # DATAHASH gas cost on tx type 2
-    tx1 = Transaction(
-        ty=2,
-        nonce=0,
-        data=to_hash_bytes(0),
-        to=datahash_zero_gas_measure_code_address,
-        gas_limit=3000000,
-        max_fee_per_gas=10,
-        max_priority_fee_per_gas=10,
-        access_list=[],
+
+@test_from(fork=ShardingFork)
+def test_datahash_invalid_blob_index(_: str):
```

Comment:
Also add a test where a block contains a blob transaction and then a type 2 tx, and this subsequent tx uses `DATAHASH`, it should return zero for all indexes where the first type 5 tx had blobs.

Similar test with two blob txs in the same block, and the second tx has less blobs, e.g., tx 1 has 2 blobs, tx 2 has 1 blob, DATAHASH(0x01) on tx 2 must return zero.